network-mux/src/Network/Mux/Bearer/Socket.hs

```diff
@@ -44,34 +44,59 @@ import qualified Network.Mux.Time as Mx
+  -> Tracer IO Mx.MuxTrace
   -> Socket.Socket
   -> MuxBearer IO
-socketAsMuxBearer tracer sd =
+socketAsMuxBearer sduTimeout_m tracer sd =
       Mx.MuxBearer {
         Mx.read    = readSocket,
         Mx.write   = writeSocket,
         Mx.sduSize = 12288
       }
     where
+      hdrLenght = 8
+
       readSocket :: HasCallStack => IO (Mx.MuxSDU, Time)
       readSocket = do
           traceWith tracer $ Mx.MuxTraceRecvHeaderStart
-          hbuf <- recvLen' True 8 []
+
+          -- Wait for the first part of the header without any timeout
+          h0 <- recvAtMost False hdrLenght
+
+          -- Optionally wait at most sduTimeout seconds for the complete SDU.
+          r_m <- case sduTimeout_m of
+                      Just sduTimeout -> timeout sduTimeout $ recvRem h0
+                      Nothing -> Just <$> recvRem h0
+          case r_m of
+                Nothing -> do
+                    traceWith tracer $ Mx.MuxTraceSDUReadTimeoutException
+                    throwM $ Mx.MuxError Mx.MuxSDUReadTimeout "Mux SDU Timeout" callStack
+                Just r -> return r
+
+      recvRem :: BL.ByteString -> IO (Mx.MuxSDU, Time)
+      recvRem h0 = do
+          hbuf <- recvLen' (hdrLenght - fromIntegral (BL.length h0)) [h0]
           case Mx.decodeMuxSDU hbuf of
-              Left  e      -> throwM e
-              Right header -> do
-                  traceWith tracer $ Mx.MuxTraceRecvHeaderEnd header
-                  traceWith tracer $ Mx.MuxTraceRecvPayloadStart (fromIntegral $ Mx.msLength header)
-                  blob <- recvLen' False (fromIntegral $ Mx.msLength header) []
-                  ts <- getMonotonicTime
-                  traceWith tracer (Mx.MuxTraceRecvDeltaQObservation header ts)
-                  traceWith tracer $ Mx.MuxTraceRecvPayloadEnd blob
-                  return (header {Mx.msBlob = blob}, ts)
-
-      recvLen' :: Bool -> Int64 -> [BL.ByteString] -> IO BL.ByteString
-      recvLen' _ 0 bufs = return (BL.concat $ reverse bufs)
-      recvLen' waitingOnNxtHeader l bufs = do
+               Left  e      ->  throwM e
+               Right header -> do
+                   traceWith tracer $ Mx.MuxTraceRecvHeaderEnd header
+                   traceWith tracer $ Mx.MuxTraceRecvPayloadStart (fromIntegral $ Mx.msLength header)
+                   blob <- recvLen' (fromIntegral $ Mx.msLength header) []
+
+                   ts <- getMonotonicTime
+                   traceWith tracer (Mx.MuxTraceRecvDeltaQObservation header ts)
+                   traceWith tracer $ Mx.MuxTraceRecvPayloadEnd blob
+                   return (header {Mx.msBlob = blob}, ts)
+
+      recvLen' ::  Int64 -> [BL.ByteString] -> IO BL.ByteString
+      recvLen' 0 bufs = return $ BL.concat $ reverse bufs
+      recvLen' l bufs = do
+          buf <- recvAtMost False l
+          recvLen' (l - fromIntegral (BL.length buf)) (buf : bufs)
+
+      recvAtMost :: Bool -> Int64 -> IO BL.ByteString
+      recvAtMost waitingOnNxtHeader l = do
```

Comment:
As far as I can see, this is always called with `False`. If that's correct we can simplify. But should it be being called with `True` somewhere, e.g. for the first part of the header as noted above?

Reply:
Yes, it should be called with true.